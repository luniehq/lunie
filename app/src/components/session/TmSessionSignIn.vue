<template>
  <SessionFrame :icon="`lock`">
    <TmFormStruct :submit="onSubmit" class="session-container">
      <h2 class="session-title">Sign in with account</h2>
      <div class="session-main bottom-indent">
        <TmFormGroup field-id="sign-in-name" field-label="Select Account">
          <TmField
            id="sign-in-name"
            v-model="signInAddress"
            :options="accounts"
            type="select"
            placeholder="Select accountâ€¦"
            vue-focus="vue-focus"
          />
          <TmFormMsg
            v-if="$v.signInAddress.$error && !$v.signInAddress.required"
            name="Name"
            type="required"
          />
        </TmFormGroup>
        <TmFormGroup
          :error="$v.signInPassword.$error"
          field-id="sign-in-password"
          field-label="Password"
        >
          <TmField
            id="sign-in-password"
            v-model="signInPassword"
            type="password"
          />
          <TmFormMsg
            v-if="$v.signInPassword.$error && !$v.signInPassword.required"
            name="Password"
            type="required"
          />
          <TmFormMsg
            v-if="$v.signInPassword.$error && !$v.signInPassword.minLength"
            name="Password"
            type="minLength"
            min="10"
          />
          <TmFormMsg v-if="error" type="custom" :msg="error" />
        </TmFormGroup>
        <TmFormGroup
          class="field-checkbox"
          field-id="sign-up-warning"
          field-label
        >
          <div class="field-checkbox-testnet">
            <label class="field-checkbox-label" for="select-testnet">
              <input id="select-testnet" v-model="testnet" type="checkbox" />
              This is a testnet address</label
            >
          </div>
        </TmFormGroup>
      </div>
      <div class="session-footer">
        <TmBtn value="Sign In" />
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { mapState, mapGetters } from "vuex"
import { required, minLength } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import TmFormGroup from "common/TmFormGroup"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import TmFormStruct from "common/TmFormStruct"
import SessionFrame from "common/SessionFrame"
import { checkAddress } from "@polkadot/util-crypto"

const isPolkadotAddress = (address) => {
  const polkadotRegexp = /^(([0-9a-zA-Z]{47})|([0-9a-zA-Z]{48}))$/
  return polkadotRegexp.test(address)
}

const isValidPolkadotAddress = (address, addressPrefix) => {
  return checkAddress(address, addressPrefix)
}

export default {
  name: `session-sign-in`,
  components: {
    TmBtn,
    TmField,
    TmFormGroup,
    TmFormMsg,
    SessionFrame,
    TmFormStruct,
  },
  data: () => ({
    signInAddress: ``,
    signInPassword: ``,
    error: ``,
    testnet: false,
  }),
  computed: {
    ...mapState([`keystore`, `session`]),
    ...mapGetters([`networks`, `network`]),
    accounts() {
      let accounts = this.keystore.accounts
      return accounts.map(({ name, address }) => ({
        value: address,
        key: name,
      }))
    },
    networkOfAddress() {
      if (isPolkadotAddress(this.signInAddress)) {
        return this.networks
          .filter(({ network_type }) => network_type === `polkadot`)
          .find(
            (network) =>
              isValidPolkadotAddress(
                this.signInAddress,
                parseInt(network.address_prefix)
              )[0]
          )
      }

      const selectedNetworksArray = this.networks.filter(({ address_prefix }) =>
        this.signInAddress.startsWith(address_prefix)
      )

      const selectedNetwork = selectedNetworksArray.find(({ testnet }) =>
        this.testnet ? testnet === true : testnet === false
      )

      return selectedNetwork
    },
  },
  created() {
    this.$store.dispatch("loadLocalAccounts")
  },
  mounted() {
    this.setDefaultAccount(this.accounts)
  },
  methods: {
    async onSubmit() {
      this.$v.$touch()
      if (this.$v.$error) return

      if (!this.networkOfAddress) {
        this.error = `No ${
          this.testnet ? "testnet" : "mainnet"
        } for this address found`
        return
      }
      const sessionCorrect = await this.$store.dispatch(`testLogin`, {
        password: this.signInPassword,
        address: this.signInAddress,
      })
      if (sessionCorrect) {
        this.selectNetworkByAddress(this.signInAddress)

        this.$store.dispatch(`signIn`, {
          password: this.signInPassword,
          address: this.signInAddress,
          sessionType: "local",
          networkId: this.network,
        })
        localStorage.setItem(`prevAccountKey`, this.signInAddress)
        this.$router.push({
          name: "portfolio",
          params: {
            networkId: this.networkOfAddress.slug,
          },
        })
      } else {
        this.error = `The provided username or password is wrong.`
      }
    },
    setDefaultAccount() {
      const prevAccountKey = localStorage.getItem(`prevAccountKey`)
      const prevAccountExists = this.accounts.find(
        (a) => a.value === prevAccountKey
      )

      if (this.accounts.length === 1) {
        this.signInAddress = this.accounts[0].value
      } else if (prevAccountExists) {
        this.signInAddress = prevAccountKey
      }

      if (this.signInAddress) {
        this.$el.querySelector(`#sign-in-password`).focus()
      } else {
        this.$el.querySelector(`#sign-in-name`).focus()
      }
    },
    async selectNetworkByAddress(address) {
      let selectedNetworksArray = this.networks.filter(({ address_prefix }) =>
        address.startsWith(address_prefix)
      )
      let selectedNetwork = ``

      // handling when there are both mainnet and testnet networks
      if (selectedNetworksArray.length > 1) {
        /* istanbul ignore next */
        selectedNetwork = selectedNetworksArray.filter(({ testnet }) =>
          this.testnet ? testnet === true : testnet === false
        )[0]
      } else {
        selectedNetwork = selectedNetworksArray[0]
      }
      if (isPolkadotAddress(this.signInAddress)) {
        selectedNetwork = this.networks
          .filter(({ network_type }) => network_type === `polkadot`)
          .find(
            (network) =>
              isValidPolkadotAddress(
                this.signInAddress,
                parseInt(network.address_prefix)
              )[0]
          )
      }

      this.$store.dispatch(`setNetwork`, selectedNetwork)
    },
  },
  validations() {
    return {
      signInAddress: { required },
      signInPassword: { required, minLength: minLength(10) },
    }
  },
}
</script>
