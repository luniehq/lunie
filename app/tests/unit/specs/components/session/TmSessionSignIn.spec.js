import { shallowMount, createLocalVue } from "@vue/test-utils"
import Vuelidate from "vuelidate"
import TmSessionSignIn from "session/TmSessionSignIn"

describe(`TmSessionSignIn`, () => {
  const localVue = createLocalVue()
  localVue.use(Vuelidate)

  let wrapper, $store

  const networks = [
    {
      id: "gaia-testnet",
      address_prefix: "cosmos",
      testnet: true,
      slug: "cosmos-testnet",
    },
    {
      id: "cosmoshub",
      address_prefix: "cosmos",
      testnet: false,
      slug: "cosmos",
    },
    {
      id: "terra-testnet",
      address_prefix: "terra",
      testnet: true,
      slug: "terra-testnet",
    },
    {
      id: "kusama",
      address_prefix: "2",
      testnet: false,
      slug: "kusama",
      network_type: "polkadot",
    },
  ]

  const addresses = [
    `cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9`,
    `cosmos1pxdf0lvq5jvl9uxznklgc5gxuwzpdy5ynem546`,
  ]

  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(() => true),
      state: {
        keys: [
          {
            name: `cosmosdefault`,
            password: `1234567890`,
            address: addresses[0],
          },
          {
            name: `terradefault`,
            password: `1234567890`,
            address: addresses[0],
          },
        ],
        session: {
          address: ``,
          addresses: [
            {
              address: `cosmos1z8mzakma7vnaajysmtkwt4wgjqr2m84tzvyfkz`,
              type: `explore`,
            },
            {
              address: `cosmos1unc788q8md2jymsns24eyhua58palg5kc7cstv`,
              type: `ledger`,
            },
            {
              address: `cosmos1vxkye0mpdtjhzrc6va5lcnxnuaa7m64khj8klc`,
              type: `extension`,
            },
            {
              address: `cosmos1vxkye0mpdtjhzrc6va5lcnxnuaa7m64khj8xyz`,
              type: `local`,
            },
          ],
        },
        keystore: {
          accounts: [
            {
              address: `cosmos1234`,
              name: `my_account`,
            },
          ],
        },
      },
      getters: {
        networks,
        network: `cosmos-hub-mainnet`,
      },
    }

    wrapper = shallowMount(TmSessionSignIn, {
      localVue,
      mocks: {
        $router: {
          push: jest.fn(),
        },
        $store,
      },
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should close the modal on successful login`, async () => {
    wrapper.setData({
      signInPassword: `1234567890`,
      signInAddress: `cosmosdefault`,
    })
    wrapper.vm.$emit = jest.fn()
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith({
      name: "portfolio",
      params: { networkId: "cosmos" },
    })
  })

  it(`should signal signedin state on successful login`, async () => {
    wrapper.setData({
      signInPassword: `1234567890`,
      signInAddress: `cosmosdefault`,
    })
    await wrapper.vm.onSubmit()
    expect($store.dispatch).toHaveBeenCalledWith(`signIn`, {
      password: `1234567890`,
      address: "cosmosdefault",
      sessionType: `local`,
      networkId: `cosmos-hub-mainnet`,
    })
  })

  it(`should show error if password not 10 long`, () => {
    wrapper.setData({ signInPassword: `123` })
    wrapper.vm.onSubmit()
    expect($store.commit.mock.calls[1]).toBeUndefined()
    expect(wrapper.find(`.tm-form-msg-error`)).toBeDefined()
  })

  it(`should show a notification if signin failed`, async () => {
    $store.dispatch = jest.fn().mockResolvedValueOnce(false)
    wrapper.setData({
      signInPassword: `1234567889`,
      signInAddress: `cosmosdefault`,
    })
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.error).toBe(`The provided username or password is wrong.`)
  })

  it(`should show the only account that exists`, () => {
    const self = {
      accounts: [
        {
          value: `default`,
        },
      ],
      $el: {
        querySelector: jest.fn(() => ({
          focus: jest.fn(),
        })),
      },
    }
    TmSessionSignIn.methods.setDefaultAccount.call(self)

    expect(self.signInAddress).toBe(`default`)
    expect(self.$el.querySelector).toHaveBeenCalledWith(`#sign-in-password`)
  })

  it(`should show the last account used`, () => {
    localStorage.setItem(`prevAccountKey`, `lastUsed`)

    const self = {
      accounts: [
        {
          value: `default`,
        },
        {
          value: `lastUsed`,
        },
      ],
      $el: {
        querySelector: jest.fn(() => ({
          focus: jest.fn(),
        })),
      },
    }
    TmSessionSignIn.methods.setDefaultAccount.call(self)

    expect(self.signInAddress).toBe(`lastUsed`)
    expect(self.$el.querySelector).toHaveBeenCalledWith(`#sign-in-password`)
  })

  it(`should focus on the name input when there are no accounts`, () => {
    const self = {
      accounts: [],
      $el: {
        querySelector: jest.fn(() => ({
          focus: jest.fn(),
        })),
      },
    }
    TmSessionSignIn.methods.setDefaultAccount.call(self)

    expect(self.signInAddress).toBe(undefined)
    expect(self.$el.querySelector).toHaveBeenCalledWith(`#sign-in-name`)
  })

  it(`automatically connects to the network an address belongs to`, async () => {
    await wrapper.vm.selectNetworkByAddress(`terradefault`)
    expect($store.dispatch).toHaveBeenCalledWith(`setNetwork`, {
      id: "terra-testnet",
      address_prefix: "terra",
      testnet: true,
      slug: "terra-testnet",
    })
  })

  it(`automatically connects to the testnet network an address belongs to if "tesnet" is set to true`, async () => {
    wrapper.setData({
      signInAddress: `HaT6pivXZTGWXM5xRBPgFKPDAKhJp2vnyUu3tcvvioHeqdt`,
    })
    await wrapper.vm.selectNetworkByAddress(
      `HaT6pivXZTGWXM5xRBPgFKPDAKhJp2vnyUu3tcvvioHeqdt`
    )
    expect($store.dispatch).toHaveBeenCalledWith(`setNetwork`, {
      id: "kusama",
      slug: "kusama",
      address_prefix: "2",
      testnet: false,
      network_type: "polkadot",
    })
  })

  it(`returns the Polkadot network we want to connect to`, async () => {
    wrapper.setData({
      testnet: true,
    })
    await wrapper.vm.selectNetworkByAddress(`cosmosdefault`)
    expect($store.dispatch).toHaveBeenCalledWith(`setNetwork`, {
      id: "gaia-testnet",
      slug: "cosmos-testnet",
      address_prefix: "cosmos",
      testnet: true,
    })
  })

  it(`checks that the address is a valid address of the network and selects testnet if testnet is set to true`, () => {
    const self = {
      testnet: true,
      signInAddress: addresses[0],
      networks,
    }
    const signInNetwork = TmSessionSignIn.computed.networkOfAddress.call(self)
    expect(signInNetwork).toEqual({
      address_prefix: "cosmos",
      id: "gaia-testnet",
      slug: "cosmos-testnet",
      testnet: true,
    })
  })

  it(`Polkadot network specifig: checks that the address is a valid address of the network 
    and selects testnet if testnet is set to true`, () => {
    const self = {
      testnet: false,
      signInAddress: `HaT6pivXZTGWXM5xRBPgFKPDAKhJp2vnyUu3tcvvioHeqdt`,
      networks,
    }
    const signInNetwork = TmSessionSignIn.computed.networkOfAddress.call(self)
    expect(signInNetwork).toEqual({
      address_prefix: "2",
      id: "kusama",
      network_type: "polkadot",
      slug: "kusama",
      testnet: false,
    })
  })

  it(`displays an error message if networkOfAddress is false`, async () => {
    wrapper.setData({
      signInAddress: `terradefault`,
      signInPassword: `1234567890`,
      testnet: false,
      error: ``,
    })
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.error).toBe(`No mainnet for this address found`)
  })
})
