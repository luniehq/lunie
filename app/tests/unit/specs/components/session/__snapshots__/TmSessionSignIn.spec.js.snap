// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionSignIn has the expected html structure 1`] = `
<sessionframe-stub
  icon="lock"
>
  <tmformstruct-stub
    class="session-container"
    submit="function () { [native code] }"
    width=""
  >
    <h2
      class="session-title"
    >
      Sign in with account
    </h2>
     
    <div
      class="session-main bottom-indent"
    >
      <tmformgroup-stub
        fieldid="sign-in-name"
        fieldlabel="Select Account"
      >
        <tmfield-stub
          id="sign-in-name"
          options="[object Object]"
          placeholder="Select accountâ€¦"
          type="select"
          value="cosmos1234"
          vue-focus="vue-focus"
        />
         
        <!---->
      </tmformgroup-stub>
       
      <tmformgroup-stub
        fieldid="sign-in-password"
        fieldlabel="Password"
      >
        <tmfield-stub
          id="sign-in-password"
          type="password"
          value=""
        />
         
        <!---->
         
        <!---->
         
        <!---->
      </tmformgroup-stub>
       
      <tmformgroup-stub
        class="field-checkbox"
        fieldid="sign-up-warning"
        fieldlabel=""
      >
        <div
          class="field-checkbox-testnet"
        >
          <label
            class="field-checkbox-label"
            for="select-testnet"
          >
            <input
              id="select-testnet"
              type="checkbox"
            />
            
            This is a testnet address
          </label>
        </div>
      </tmformgroup-stub>
    </div>
     
    <div
      class="session-footer"
    >
      <tmbtn-stub
        customclass=""
        value="Sign In"
      />
    </div>
  </tmformstruct-stub>
</sessionframe-stub>
`;
